File: sdk/go/transport/legacy.go
package transport

import (
	"context"

	latent "latent.dev/sdk/go"
	"latent.dev/sdk/go/profile"
)

type LegacyClient struct {
	client *Client
}

var _ latent.Client = (*LegacyClient)(nil)

func NewLegacy(ctx context.Context, config Config) (*LegacyClient, error) {
	client, failure := New(ctx, config)
	if failure != nil {
		return nil, failure
	}
	return client.Legacy(), nil
}

func (client *Client) Legacy() *LegacyClient { return &LegacyClient{client: client} }

func (client *LegacyClient) Close() error { return client.client.Close() }

func (client *LegacyClient) Invoke(ctx context.Context, request latent.InvokeRequest) (latent.InvocationOutcome, error) {
	budget := request.Options.Budget
	response, failure := client.client.Invoke(ctx, profile.InvokeRequest{
		ActivationId: request.ActivationID, RootActivationId: request.RootActivationID, ParentActivationId: request.ParentActivationID,
		Target: &profile.InvocationTarget{Tenant: request.Target.Tenant, Service: request.Target.Service,
			Contract: request.Target.Contract, Function: request.Target.Function, Route: request.Target.Route},
		Payload: request.Payload, MediaType: request.MediaType, DeadlineUnixMillis: request.Options.DeadlineUnixMillis,
		Priority: uint32(request.Options.Priority), IdempotencyKey: request.Options.IdempotencyKey, Metadata: request.Options.Metadata,
		Budget: &profile.ResourceBudget{
			CpuFuel: budget.CPUFuel, MemoryBytes: budget.MemoryBytes, WallTimeLimitMillis: budget.WallTimeLimitMillis,
			ChildCalls: budget.ChildCalls, OutboundRequests: budget.OutboundRequests,
			StateReadBytes: budget.StateReadBytes, StateWriteBytes: budget.StateWriteBytes,
			BlobReadBytes: budget.BlobReadBytes, BlobWriteBytes: budget.BlobWriteBytes,
			LogBytes: budget.LogBytes, EffectCount: budget.EffectCount,
		},
	}, profile.CallOptions{})
	if failure != nil {
		return latent.InvocationOutcome{}, failure
	}
	value := response.Value
	receipt := latent.InvocationReceipt{ActivationID: value.ActivationId, RevisionID: value.RevisionId,
		ReleaseDigest: value.ReleaseDigest, PublicationID: value.PublicationId, RouteGeneration: value.RouteGeneration,
		Consumption: legacyConsumption(*value.Consumption)}
	if value.Success != nil {
		return latent.InvocationOutcome{Success: &latent.InvokeResponse{
			ActivationID: receipt.ActivationID, RevisionID: receipt.RevisionID, ReleaseDigest: receipt.ReleaseDigest,
			PublicationID: receipt.PublicationID, RouteGeneration: receipt.RouteGeneration, Consumption: receipt.Consumption,
			Payload: value.Success.Payload, MediaType: value.Success.MediaType, CommittedStateVersion: value.Success.CommittedStateVersion,
			EffectIDs: value.Success.EffectIds, Metadata: value.Success.Metadata,
		}}, nil
	}
	if value.DeclaredError != nil {
		return latent.InvocationOutcome{DeclaredError: &latent.DeclaredInvocationError{Receipt: receipt, Error: legacyDeclared(*value.DeclaredError)}}, nil
	}
	return latent.InvocationOutcome{PlatformFailure: &latent.PlatformInvocationFailure{Receipt: receipt, Error: legacyPlatform(*value.PlatformFailure)}}, nil
}

func (client *LegacyClient) Cancel(ctx context.Context, activationID, reason string) (latent.CancelResponse, error) {
	response, failure := client.client.Cancel(ctx, profile.CancelRequest{ActivationId: activationID, Reason: reason}, profile.CallOptions{})
	if failure != nil {
		return latent.CancelResponse{}, failure
	}
	dispositions := map[profile.CancelDisposition]latent.CancelDisposition{
		profile.CancelDispositionAccepted:        latent.CancelAccepted,
		profile.CancelDispositionAlreadyTerminal: latent.CancelAlreadyTerminal,
		profile.CancelDispositionNotFound:        latent.CancelNotFound,
	}
	return latent.CancelResponse{Disposition: dispositions[response.Value.Disposition], TerminalState: response.Value.TerminalState}, nil
}

func (client *LegacyClient) GetActivation(ctx context.Context, activationID string) (latent.ActivationStatus, error) {
	response, failure := client.client.GetActivation(ctx, profile.GetActivationRequest{ActivationId: activationID}, profile.CallOptions{})
	if failure != nil {
		return latent.ActivationStatus{}, failure
	}
	value := response.Value
	result := latent.ActivationStatus{ActivationID: value.ActivationId, Phase: value.Phase,
		TerminalState: value.TerminalState, LastUpdatedUnixMS: value.LastUpdatedUnixMillis,
		TerminalAtUnixMS: value.TerminalAtUnixMillis, Metadata: value.Metadata}
	if value.FinalConsumption != nil {
		consumption := legacyConsumption(*value.FinalConsumption)
		result.FinalConsumption = &consumption
	}
	if value.Succeeded != nil {
		result.TerminalOutcome = &latent.RetainedInvocationOutcome{Succeeded: &latent.ActivationSuccessSummary{
			CommittedStateVersion: value.Succeeded.CommittedStateVersion, EffectIDs: value.Succeeded.EffectIds, Metadata: value.Succeeded.Metadata}}
	} else if value.DeclaredError != nil {
		declared := legacyDeclared(*value.DeclaredError)
		result.TerminalOutcome = &latent.RetainedInvocationOutcome{DeclaredError: &declared}
	} else if value.PlatformFailure != nil {
		platform := legacyPlatform(*value.PlatformFailure)
		result.TerminalOutcome = &latent.RetainedInvocationOutcome{PlatformFailure: &platform}
	}
	return result, nil
}

func legacyConsumption(value profile.BudgetConsumption) latent.BudgetConsumption {
	return latent.BudgetConsumption{CPUFuel: value.CpuFuel, PeakMemoryBytes: value.PeakMemoryBytes,
		WallTimeMicros: value.WallTimeMicros, ChildCalls: value.ChildCalls, OutboundRequests: value.OutboundRequests,
		StateReadBytes: value.StateReadBytes, StateWriteBytes: value.StateWriteBytes,
		BlobReadBytes: value.BlobReadBytes, BlobWriteBytes: value.BlobWriteBytes, LogBytes: value.LogBytes, EffectCount: value.EffectCount}
}

func legacyDeclared(value profile.DeclaredError) latent.DeclaredError {
	return latent.DeclaredError{Code: value.Code, Message: value.Message, Payload: value.Payload, MediaType: value.MediaType, Metadata: value.Metadata}
}

func legacyPlatform(value profile.PlatformError) latent.PlatformError {
	details := make([]latent.ErrorDetail, len(value.DetailItems))
	for index, detail := range value.DetailItems {
		details[index] = latent.ErrorDetail{Kind: detail.Kind, Fields: detail.Fields}
	}
	return latent.PlatformError{Code: value.Code, Message: value.Message, Retryable: value.Retryable, Details: details}
}

File: sdk/java-client/src/transport/java/dev/latent/sdk/transport/Legacy.java
package dev.latent.sdk.transport;

import dev.latent.sdk.Management;
import dev.latent.sdk.Models;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

final class Legacy {
    private static final Management.CallOptions DEFAULT = new Management.CallOptions(Optional.empty());
    private Legacy() { }

    static <Source, Target> CompletableFuture<Target> mapped(CompletableFuture<Source> source, Function<Source, Target> convert) {
        var target = new CompletableFuture<Target>() {
            @Override public boolean cancel(boolean mayInterruptIfRunning) { return source.cancel(mayInterruptIfRunning); }
        };
        source.whenComplete((value, error) -> {
            if (error != null) target.completeExceptionally(error);
            else {
                try { target.complete(convert.apply(value)); }
                catch (RuntimeException failure) { target.completeExceptionally(failure); }
            }
        });
        return target;
    }

    static CompletableFuture<Models.InvocationOutcome> invoke(RpcClient client, Models.InvokeRequest request) {
        var target = request.target();
        var options = request.options();
        var budget = options.budget();
        var portable = new Management.InvokeRequest(request.activationId(), request.parentActivationId(), request.rootActivationId(),
                Optional.of(new Management.InvocationTarget(target.tenant(), target.service(), target.contract(), target.function(), target.route())),
                request.payload(), request.mediaType(), options.deadlineUnixMillis(), Byte.toUnsignedInt(options.priority()), options.idempotencyKey(),
                Optional.of(new Management.ResourceBudget(budget.cpuFuel(), budget.memoryBytes(), budget.childCalls(), budget.outboundRequests(),
                        budget.stateReadBytes(), budget.stateWriteBytes(), budget.blobReadBytes(), budget.blobWriteBytes(), budget.logBytes(),
                        budget.effectCount(), budget.wallTimeLimitMillis())), options.metadata());
        return mapped(client.invoke(portable, DEFAULT), response -> outcome(response.value()));
    }

    static Models.BudgetConsumption consumption(Management.BudgetConsumption value) {
        return new Models.BudgetConsumption(value.cpuFuel(), value.peakMemoryBytes(), value.wallTimeMicros(), value.childCalls(),
                value.outboundRequests(), value.stateReadBytes(), value.stateWriteBytes(), value.blobReadBytes(), value.blobWriteBytes(),
                value.logBytes(), value.effectCount());
    }

    static Models.DeclaredError declared(Management.DeclaredError value) {
        return new Models.DeclaredError(value.code(), value.message(), value.payload(), value.mediaType(), value.metadata());
    }

    static Models.PlatformFailure platform(Management.PlatformError value) {
        return new Models.PlatformFailure(value.code(), value.message(), value.retryable(), value.detailItems().stream()
                .map(detail -> new Models.ErrorDetail(detail.kind(), detail.fields())).toList());
    }

    static Models.InvocationOutcome outcome(Management.InvokeResponse response) {
        var consumption = consumption(response.consumption().orElseThrow());
        if (response.success().isPresent()) {
            var result = response.success().get();
            return new Models.InvocationSuccess(new Models.InvokeResponse(response.activationId(), response.revisionId(), response.releaseDigest(),
                    response.routeGeneration(), result.payload(), result.mediaType(), result.committedStateVersion(), result.effectIds(), consumption,
                    result.metadata(), response.publicationId()));
        }
        var receipt = new Models.InvocationReceipt(response.activationId(), response.revisionId(), response.releaseDigest(), response.routeGeneration(),
                consumption, response.publicationId());
        if (response.declaredError().isPresent()) return new Models.DeclaredInvocationError(receipt, declared(response.declaredError().get()));
        return new Models.PlatformInvocationFailure(receipt, platform(response.platformFailure().orElseThrow()));
    }

    static CompletableFuture<Models.CancelResponse> cancel(RpcClient client, String identity, String reason) {
        return mapped(client.cancel(new Management.CancelRequest(identity, reason), DEFAULT), response -> {
            var value = response.value();
            Models.CancelDisposition disposition = switch (value.disposition().value()) {
                case 1 -> Models.CancelDisposition.ACCEPTED;
                case 2 -> Models.CancelDisposition.ALREADY_TERMINAL;
                case 3 -> Models.CancelDisposition.NOT_FOUND;
                default -> throw new Management.ClientException(new Management.ClientFailure(Management.FailureCategory.DECODE,
                        "unsupported cancellation disposition", Optional.empty(), Optional.empty(), true, response.metadata().outcome(),
                        response.metadata().identity(), response.metadata().auditAck(), response.metadata().auditStatus(),
                        Optional.of(new Management.UnsupportedWireValue("cancel.disposition", Integer.toString(value.disposition().value()))),
                        response.metadata().auditAttemptSequence()));
            };
            return new Models.CancelResponse(disposition, value.terminalState());
        });
    }

    static CompletableFuture<Models.ActivationStatus> getActivation(RpcClient client, String identity) {
        return mapped(client.getActivation(new Management.GetActivationRequest(identity), DEFAULT), response -> {
            var value = response.value();
            Optional<Models.RetainedInvocationOutcome> outcome = Optional.empty();
            if (value.succeeded().isPresent()) {
                var success = value.succeeded().get();
                outcome = Optional.of(new Models.ActivationSuccessSummary(success.committedStateVersion(), success.effectIds(), success.metadata()));
            } else if (value.declaredError().isPresent()) outcome = Optional.of(new Models.RetainedDeclaredError(declared(value.declaredError().get())));
            else if (value.platformFailure().isPresent()) outcome = Optional.of(new Models.RetainedPlatformFailure(platform(value.platformFailure().get())));
            return new Models.ActivationStatus(value.activationId(), value.phase(), value.terminalState(), outcome,
                    value.finalConsumption().map(Legacy::consumption), value.lastUpdatedUnixMillis(), value.terminalAtUnixMillis(), value.metadata());
        });
    }
}
